Feat/auto restart connectors (#7500)
strimzi/strimzi-kafka-operator: documentation/api/io.strimzi.api.kafka.model.AutoRestart.adoc

Connectors and tasks that are in the FAILED state are automatically restarted.

A back-off algorithm is applied to automatic restarts of each connector and it's tasks.

> When a connector is failing an auto restart attempt will happen at first and second reconcile then attempts while be delayed

Reviewer: ```suggestion
When a connector is failing an auto restart attempt will happen at first and second reconcile then attempts will be delayed,
```

incrementing the time between automatic restarts by 2 minutes up to a maximum of 30 minutes, at which point no further

> automatic restarts are attempted, and the connector or tasks are left in the FAILED state.

Reviewer: Should we add a sentence here to explain what the user should do next? Is the expectation that they then would add an annotation to request the connector be restarted? Once the connector/task has been successfully restarted will automatic restarts work again if it later fails?

Author: Yes I missed parts of what user should expect after auto restart and what should they do next. I will add them.

> If you are using a kafka version <3.0.0, only the connector will be restarted not failed tasks.

Reviewer: ```suggestion
If you are using a Kafka version <3.0.0, only the connector will be restarted not failed tasks.
```

To disable the automatic restarting of the connector and it's tasks when they are in the FAILED state on a KafkaConnector resource:

[source,yaml,subs="attributes+"]
----
apiVersion: strimzi.io/v1alpha1
kind: KafkaConnector
metadata:
    name: my-source-connector
spec:
    autoRestart:
        enabled: false
----

> Similarly, a new field in the KafkaMirrorMaker2 spec named autoRestartConnectorsAndTasks with an enabled field value of false can be applied, to disable auto-restarting connectors and tasks.

Reviewer 2: ```suggestion
For MirrorMaker 2.0, use the `autoRestartConnectorsAndTasks` property of the `KafkaMirrorMaker2` resource to disable automatic restarts of failed connectors and tasks.
```

> [source,yaml,subs="attributes+"]

Reviewer 2: ```suggestion
.Disabling automatic restarts of failed connectors for MirrorMaker 2.0
[source,yaml,subs="attributes+"]
```

----
apiVersion: strimzi.io/v1alpha1
kind: KafkaMirrorMaker2
metadata:
  name: my-mm2-cluster
spec:
  autoRestartConnectorsAndTasks:

>     enabled: false

Reviewer 3: As mentioned above, this matchs the proposal. But not your current implementation.

The connector and tasks remain in a `FAILED` state.
It means that you will have to restart connector/tasks manually, for example by annotating KafKaConnector CR with `strimzi.io/restart: "true"`

If you are using a Kafka version earlier than 3.0.0, only the failed connector is restarted not failed tasks.

For Kafka Connect connectors, use the `autoRestart` property of the `KafkaConnector` resource to enable automatic restarts of failed connectors and tasks.

.Enabling automatic restarts of failed connectors for Kafka Connect
[source,yaml,subs="attributes+"]
----
apiVersion: strimzi.io/v1alpha1
kind: KafkaConnector
metadata:
    name: my-source-connector
spec:
    autoRestart:
        enabled: true
----

For MirrorMaker 2.0, use the `autoRestartConnectorsAndTasks` property of the `KafkaMirrorMaker2` resource to enable automatic restarts of failed connectors and tasks.

.Enabling automatic restarts of failed connectors for MirrorMaker 2.0
[source,yaml,subs="attributes+"]
----
apiVersion: strimzi.io/v1alpha1
kind: KafkaMirrorMaker2
metadata:
  name: my-mm2-cluster
spec:
  autoRestart:

>     enabled: true

Reviewer 3: This should be updated to match the current API layout with per-connector configurations I guess?
